src/pretix/plugins/stripe/payment.py

```diff
@@ -110,41 +110,40 @@ def payment_perform(self, request, order) -> str:
             else:
                 err = {'message': str(e)}
                 logger.exception('Stripe error: %s' % str(e))
-            messages.error(request, _('Stripe reported an error with your card: %s' % err['message']))
             logger.info('Stripe card error: %s' % str(err))
             order.payment_info = json.dumps({
                 'error': True,
                 'message': err['message'],
             })
             order.save()
+            raise PaymentException(_('Stripe reported an error with your card: %s' % err['message']))
         except stripe.error.StripeError as e:
             if e.json_body:
                 err = e.json_body['error']
                 logger.exception('Stripe error: %s' % str(err))
             else:
                 err = {'message': str(e)}
                 logger.exception('Stripe error: %s' % str(e))
-            messages.error(request, _('We had trouble communicating with Stripe. Please try again and get in touch '
-                                      'with us if this problem persists.'))
             order.payment_info = json.dumps({
                 'error': True,
                 'message': err['message'],
             })
             order.save()
+            raise PaymentException(_('We had trouble communicating with Stripe. Please try again and get in touch '
+                                      'with us if this problem persists.'))
         else:
             if charge.status == 'succeeded' and charge.paid:
                 try:
                     mark_order_paid(order, 'stripe', str(charge))
                 except Quota.QuotaExceededException as e:
-                    messages.error(request, str(e))
+                    raise PaymentException(e)
                 except SendMailException:
-                    messages.warning(request, _('There was an error sending the confirmation mail.'))
-
+                    raise PaymentException(_('There was an error sending the confirmation mail.'))
```

Comment:
We talked about that this promotes a "warning" to an "error". I thought about it, and I believe it's fine.